mtraining/src/main/java/org/motechproject/mtraining/service/impl/CourseStructureServiceImpl.java

```diff
@@ -0,0 +1,231 @@
+import org.motechproject.mtraining.service.CourseStructureService;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Service;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Implementation of the {@link org.motechproject.mtraining.service.CourseStructureService}.
+ * This service allows creating relations between course units(Course, Chapter, Quiz and Lesson).
+ */
+@Service("courseStructureService")
+public class CourseStructureServiceImpl implements CourseStructureService {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(CourseStructureServiceImpl.class);
+
+    @Autowired
+    private CourseDataService courseDataService;
+
+    @Autowired
+    private ChapterDataService chapterDataService;
+
+    @Autowired
+    private LessonDataService lessonDataService;
+
+    @Autowired
+    private QuizDataService quizDataService;
+
+    @Override
+    public void updateCourseStructure(List<CourseUnitDto> courses) {
+        // maps to store objects from old course structure
+        Map<Long, Course> courseMap = new HashMap<>();
+        Map<Long, Chapter> chapterMap = new HashMap<>();
+        Map<Long, Lesson> lessonMap = new HashMap<>();
+        Map<Long, Quiz> quizMap = new HashMap<>();
+
+        //we clear old relations and fill maps with objects to update
+        clearRelationships(courses, courseMap, chapterMap, lessonMap, quizMap);
+
+        for (CourseUnitDto courseUnitDto : courses) {
+            LOGGER.debug("Start processing course with id {} and name {}", courseUnitDto.getId(), courseUnitDto.getName());
+
+            Course course = courseMap.get(courseUnitDto.getId());
+
+            if (course == null) {
+                throw new CourseUnitNotFoundException("Cannot find course with id: " + courseUnitDto.getId());
+            }
+
+            List<Chapter> chapters = new ArrayList<>();
+            if (courseUnitDto.getUnits() != null) {
+                for (CourseUnitDto chapterUnitDto : courseUnitDto.getUnits()) {
+                    Chapter chapter = getChapter(chapterUnitDto.getId(), chapterMap);
+
+                    setUpLessons(chapterUnitDto, chapter, lessonMap);
+                    setUpQuiz((ChapterUnitDto) chapterUnitDto, chapter, quizMap);
+
+                    chapter.setCourse(course);
+                    chapter.setState(CourseUnitState.valueOf(chapterUnitDto.getState()));
+                    chapters.add(chapter);
+                    chapterMap.remove(chapterUnitDto.getId());
+                }
+            }
+
+            course.setState(CourseUnitState.valueOf(courseUnitDto.getState()));
+            course.setChapters(chapters);
+            courseMap.remove(courseUnitDto.getId());
+            // we must evict cache to avoid relations inconsistent
+            courseDataService.evictAllCache();
```

Comment:
Didn't you mean to use evictEntityCache here and below? EvictAllCache clears the entire Datanucleus cache for all entities